src/qibolab/transpilers/placer.py

```diff
@@ -0,0 +1,261 @@
+        example [1,2,0] or {"q0":1, "q1":2, "q2":0} to assign the
+        physical qubits 0;1;2 to the logical qubits 1;2;0 respectively.
+        connectivity (networkx.Graph): chip connectivity.
+        verbose (Bool): if "True" print info messages.
+        """
+        self.connectivity = connectivity
+        self.map = map
+        self.verbose = verbose
+
+    def __call__(self, circuit=None):
+        """Return the custom placement if it can be applied to the given circuit (if given).
+
+        Args:
+            circuit (qibo.models.Circuit): Circuit to be transpiled.
+        """
+        if isinstance(self.map, dict):
+            pass
+        elif isinstance(self.map, list):
+            self.map = dict(zip(list("q" + str(i) for i in range(len(self.map))), self.map))
+        else:
+            raise_error(TypeError, "Use dict or list to define mapping.")
+        if circuit is not None:
+            if not check_placement(circuit, self.map, self.verbose):
+                raise_error(ValueError)
+        elif not check_mapping_consistency(self.map, self.verbose):
+            raise_error(ValueError)
+        return self.map
+
+
+class Subgraph(Placer):
+    """
+    Subgraph isomorphism qubit placer, NP-complete it can take a long time
+    for large circuits. This initialization method may fail for very short circuits.
+
+    Attributes:
+        connectivity (networkx.Graph): chip connectivity.
+    """
+
+    def __init__(self, connectivity):
+        """Args:
+        connectivity (networkx.graph): chip connectivity.
+        """
+        self.connectivity = connectivity
+
+    def __call__(self, circuit: Circuit):
+        """Find the initial layout of the given circuit using subgraph isomorphism.
+
+        Args:
+            circuit (qibo.models.Circuit): Circuit to be transpiled.
+        """
+        # TODO fix networkx.GM.mapping for small subgraphs
+        circuit_repr = create_circuit_repr(circuit)
+        if len(circuit_repr) < 3:
+            raise_error(ValueError, "Circuit must contain at least two two qubit gates to implement subgraph placement")
+        H = nx.Graph()
+        H.add_nodes_from([i for i in range(self.connectivity.number_of_nodes())])
+        GM = nx.algorithms.isomorphism.GraphMatcher(self.connectivity, H)
+        i = 0
+        H.add_edge(circuit_repr[i][0], circuit_repr[i][1])
+        while GM.subgraph_is_monomorphic() == True:
+            result = GM
+            i += 1
+            H.add_edge(circuit_repr[i][0], circuit_repr[i][1])
+            GM = nx.algorithms.isomorphism.GraphMatcher(self.connectivity, H)
+            if self.connectivity.number_of_edges() == H.number_of_edges() or i == len(circuit_repr) - 1:
+                keys = list(result.mapping.keys())
+                keys.sort()
+                return {i: result.mapping[i] for i in keys}
+        keys = list(result.mapping.keys())
+        keys.sort()
+        return {i: result.mapping[i] for i in keys}
```

Comment:
```suggestion
        return dict(sorted(result.mapping.items()))
```

Reply:
here keys have the "q"

Reply:
This I did not understand:

- you're making a `dict` anyhow, so dictionary comprehension or constructing from tuples is equivalent
- in particular, your comprehension here is something like `{i: d[i] for i in d.keys()}`
- the former is not fully true, because otherwise it would be just the identity, but the difference is only that keys are sorted
- since tuples sorting is lexicographical (compare first elements first, then second, then ...), and your keys are all different, sorting `(key, value)` tuples results in the same sorting as the one on keys (since `value` is never taken into account).

To break down in steps:

```py
keys = list(result.mapping.keys())
```
is equivalent to:
```py
keys = list(result.mapping)
```

```py
keys = list(result.mapping)
keys.sort()
```
is equivalent to:
```py
keys = sorted(result.mapping)
```

```py
{i: result.mapping[i] for i in sorted(result.mapping)}
```
is equivalent to:
```py
dict(sorted(result.mapping.items()))
```
with the assumption that there are no keys considered equal (and since keys are all different because it's a `dict`, and `str`s, as `int`s, are `==` iff they are the exact same, that's a very reasonable assumption).

Reply:
In general, every time that you're iterating over the keys, and then accessing the values, i.e. something like:
```py
for k in mydict:  # or `mydict.keys()`, that's equivalent
    f(k, mydict[k])
```
there is almost always a way to rephrase as an iteration over `mydict.items()`.